File: queue/queue.c
#include <stdio.h>

typedef struct Queue {
    char *head;
    struct Queue *tail;
} Queue;

Queue newQueue() {
    static Queue temp;
    temp.head = NULL;
    temp.tail = NULL;
    return temp;
}

void push(Queue *queue, char *data) {
    if (queue->head == NULL) {
        queue->head = data;
        return;
    }
    if (queue->tail == NULL) {
        Queue ns = newQueue();
        ns.head = data;
        queue->tail = &ns;
    } else {
        push(queue->tail, data);
    }
}

char *pop(Queue *queue) {
    char *temp = queue->head;
    if (queue->tail) {
        queue->head = queue->tail->head;
        queue->tail = queue->tail->tail;
    } else {
        queue->head = NULL;
    }
    return temp;
}

int main(void) {
    Queue a = newQueue();
    push(&a, "h");
    push(&a, "w");
    printf("%p\n", pop(&a));
    printf("%p\n", pop(&a));
    printf("%s\n", pop(&a));
    return 0;
}

File: queue/_queue.rs
#[derive(Debug)]
struct Queue<T> {
    head: Option<T>,
    tail: Option<Box<Queue<T>>>,
}

impl<T> Queue<T> {
    fn new() -> Queue<T> {
        Queue {
            head: None,
            tail: None,
        }
    }

    fn push(&mut self, data: T) {
        if self.head.is_none() {
            self.head = Some(data);
            return;
        }
        match self.tail {
            Some(ref mut current_data) => current_data.push(data),
            None => {
                self.tail = Some(Box::new(Queue{
                    head: Some(data),
                    tail: None,
                }));
            }
        }
    }

    fn pop(&mut self) -> Option<T> {
        let temp = &self.head;
        match self.tail {
            Some(ref next) => {
                self.head = next.head;
                self.tail = next.tail;
            }
            None => self.head = None,
        }
        *temp
    }
}

fn main() {
    let mut a = Queue::new();
    a.push(1);
    a.push(2);
    println!("{:?}", a.pop());
    println!("{:?}", a.pop());
    println!("{:?}", a.pop());
}